solarforecastarbiter/datamodel.py

```diff
@@ -1052,6 +1055,40 @@ class ProcessedForecastObservation(BaseModel):
     observation_values: Union[pd.Series, str, None]
 
 
+@dataclass(frozen=True)
+class MetricValue(BaseModel):
+    category: str
+    metric: str
+    index: str
+    value: float
+
+
+@dataclass(frozen=True)
+class MetricResult(BaseModel):
+    name: str
+    forecast_id: str
+    observation_id: str
+    values: Tuple[MetricValue, ...]
+    reference_forecast_id: str = ''
+
+
+@dataclass(frozen=True)
+class ReportFigure(BaseModel):
+    name: str
+    div: str
+    svg: str
+    type: str
+    category: str = ''
+    metric: str = ''
+
+
+@dataclass(frozen=True)
+class RawReportPlots(BaseModel):
+    bokeh_version: str
+    script: str
+    figures: Tuple[ReportFigure, ...]
+
+
```

Comment:
what other information might we need/want to store for each metric/plot?

Reply:
sufficient for probabilistic forecasts? works for aggregates?

Reply:
Not a bad idea, especially if one wanted to only look at the same constant value for all forecasts

Reply:
Thoughts on `MetricResult`:
* Should it have a attribute for the `MetricValue`? or at least an indicator for the type of metric? Or perhaps it is handled well enough in the `Report` object I haven't gotten that far
* Would we want total number of values calculated over here?  I think it would be useful to store that.
* Also, is this where we would want to store number of dropped values, fill-in values?

I really like the report figure, but what about the plots and other types of figures that aren't going to be correlated with metrics.  I.e., should this be metric specific? `MetricFigure` and have `ReportFigure` be more generic?

Reply:
> Should it have a attribute for the MetricValue? or at least an indicator for the type of metric? Or perhaps it is handled well enough in the Report object I haven't gotten that far

``values``is a tuple of all the MetricValue for that forecast/observation pair, so I'm not quite sure what you mean.

> Would we want total number of values calculated over here? I think it would be useful to store that.

total values in the timeseries? If so, then I think that belongs with ProcessedForecastObservation

> Also, is this where we would want to store number of dropped values, fill-in values?

I think that also belongs with ProcessedForecastObservation

The class naming isn't the best, but I think results from preprocessing.py should likely be ProcessedForecastObservation objects and calculator.py should be Metrics* objects.

Reply:
> I really like the report figure, but what about the plots and other types of figures that aren't going to be correlated with metrics. I.e., should this be metric specific? MetricFigure and have ReportFigure be more generic?

That's where I see the type parameters being useful (and category and metric are not required). Maybe there a different parameter name/type could make this useful for both metric figures and other report figures.

Reply:
Makes sense. Thanks.

> That's where I see the type parameters being useful (and category and metric are not required). Maybe there a different parameter name/type could make this useful for both metric figures and other report figures.

Sounds good. I think it makes sense and could be extended later on to maybe use a kwarg dict later or use a separate dataclass object if potential other figures need to be added.